[RNMobile] Editor Onboarding: "The Basics" help section -  Topics list UI

AmandaRiu opened the pull request:
**Gutenberg-Mobile PR:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3702
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## NOTE
**This feature is only available in debug builds**


## Description
<!-- Please describe what you have changed or added -->
This PR is the first in a series of PRs to add the Editor onboarding basic help section defined in this ticket: https://github.com/WordPress/gutenberg/issues/30626 and covers the following:
- Adds the "Help" menu option to Gutenberg demo app on Android and iOS to make testing easier. 
- Adds the main help topics list `BottomSheet` and the help topic detail sub bottom sheet. Clicking an item in the help topics list will open the sub-sheet detail page with the appropriate title and back button. The actual details of the help article will be added in a later PR.
- Added additional icons to the `@wordpress/icons` package where needed to match the designs. 

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. Run the demo app
2. Open the menu on iOS by clicking the overflow menu icon in the ActionBar or by shaking the device on Android
3. Tap "Help" - the main help topics bottom sheet should be displayed. 
4. Tap one of the help topics. A new bottom sheet will be displayed. Verify the topic heading matches the topic tapped, and that the `< Back` button is available. 
5. Tap the `< Back` button to go back to the main help topics list. 
6. Tap outside the bottom sheet to dismiss. 



https://user-images.githubusercontent.com/5810477/124996585-2e43d880-e017-11eb-8fd7-cc4ae52038de.mp4



## Screenshots <!-- if applicable -->

Light | Dark
-- | --
![help-topics-android](https://user-images.githubusercontent.com/5810477/124687363-bc4a8280-dea2-11eb-96bf-5dafb6eb35d4.png)|![help-topics-dark-android](https://user-images.githubusercontent.com/5810477/124687368-beacdc80-dea2-11eb-869c-c9599f9bed57.png)
![help-topics-ios](https://user-images.githubusercontent.com/5810477/124687374-c40a2700-dea2-11eb-884d-766e1e0fb255.png)|![help-topics-dark-ios](https://user-images.githubusercontent.com/5810477/124687378-c5d3ea80-dea2-11eb-9e44-e897bbf2eb6e.png)


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature. Only available in non-production builds. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ --> 
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Commit:
Add help menu option to iOS demo app

Commit:
Add basic help topics bottom sheet

Commit:
Add help option to Android demo app

Commit:
Add new help topics icons

Commit:
Add help topics to main help bottom sheet

AmandaRiu commented:
Hi @dcalhoun 👋 This PR takes care of adding the main topics UI. I'm not sure if [how I'm currently opening the help menu](https://github.com/WordPress/gutenberg/pull/33240/files#diff-f06faf4815cadf891d1ed891241c2e2bcaa2381e8a3043115d49fbdc1043d621R162) is the best way of doing things so welcome some additional pointers in that area specifically and I can tweak it in the next round of PRs. There is still a lot of code to go but since there were already so many changes I decided to break these changes into their own PR to make it easier to review. Please let me know what you think could be changed/improved. Many thanks!

Commit:
Remove unused imports

Commit:
Optimized new SVG images using https://jakearchibald.github.io/svgomg/

Commit:
Merge branch 'trunk' into rnmobile/editor-onboarding-basic-help-2

Commit:
Add navigation to help detail subsheet

Commit:
Add react-native and react-navigation to package.json per lint

Commit:
Fix lint errors

AmandaRiu commented:
@dcalhoun I've updated this PR to add navigation to the help topic detail view as we discussed. This PR is ready for review. 👍

dcalhoun commented on packages/editor/package.json:
```diff
@@ -31,6 +31,7 @@
 	],
 	"dependencies": {
 		"@babel/runtime": "^7.13.10",
+		"@react-navigation/native": "5.7.0",
```
After considering the placement of this dependency more, I remembered an internal discussion about this. I will share a link to that via a direct message. 

We should likely either request a web reviewer help verify adding this dependency will not negatively impact web package publishing or relocate the new dependencies in this file into `packages/react-native-editor/package.json` to follow the pattern the majority of native dependencies follow today.

AmandaRiu replied:
Hrm, that's a good point. Thanks for pointing that out. I'll look into it further 👍

AmandaRiu replied:
Removed in 6cd93f575749822104966d1811005952e8e61bfb

dcalhoun commented on packages/editor/src/components/editor-help/help-detail-navigation-screen.js:
```diff
@@ -0,0 +1,37 @@
+ * WordPress dependencies
+ */
+import { BottomSheet } from '@wordpress/components';
+import { useNavigation } from '@react-navigation/native';
```
I believe we should relocated this import to the `External dependencies` comment section.

AmandaRiu replied:
Fixed in 473ec54a850f307d0f4ffc86a9c38bed0fb807c0

dcalhoun commented on packages/editor/src/components/editor-help/help-detail-navigation-screen.js:
```diff
@@ -0,0 +1,37 @@
+ * Internal dependencies
+ */
+import styles from './style.scss';
+import { __ } from '@wordpress/i18n';
```
I believe we should relocated this import to the `WordPress dependencies` comment section.

AmandaRiu replied:
Fixed in 473ec54a850f307d0f4ffc86a9c38bed0fb807c0

dcalhoun commented on packages/editor/src/components/editor-help/help-detail-navigation-screen.js:
```diff
@@ -0,0 +1,37 @@
+/**
+ * External dependencies
+ */
+import { View, Text } from 'react-native';
```
It looks like this line is currently causing linter errors. It may be related to the dependencies added to `packages/editor/package.json`. Reverting that change and relocating the relevant dependencies to `packages/react-native-editor/package.json` may resolve the error. I'm not sure.

AmandaRiu replied:
It actually ended being due to the naming of the file. I forgot to use `.native.js` as the extension.

dcalhoun commented on packages/editor/src/components/editor-help/index.native.js:
```diff
@@ -0,0 +1,92 @@
+			hasNavigation
+		>
+			<BottomSheet.NavigationContainer animate main>
+				<BottomSheet.NavigationScreen name={ 'Topics' }>
```
I believe we made need to internationalize this string. 
```suggestion
				<BottomSheet.NavigationScreen name={ __( 'Topics' ) }>
```

dcalhoun commented on packages/editor/src/components/editor-help/index.native.js:
```diff
@@ -0,0 +1,92 @@
+					<View style={ styles.bottomSheetHeader }>
+						<Text
+							style={ bottomSheetHeaderTitleStyle }
+							maxFontSizeMultiplier={ 3 }
```
Was not aware of this property. Is this done to avoid the layout issues for very large text from OS-level text scaling? Is it worth considering not limiting this and addressing those layout issues instead for improved accessibility?

AmandaRiu replied:
I found it being used heavily in any of the bottom sheet code ([bottom-sheet](https://github.com/WordPress/gutenberg/blob/5f5b838a315c8044b17781b87b4c1f6cf4fe8a78/packages/components/src/mobile/bottom-sheet/index.native.js#L477), [block-variation-picker](https://github.com/WordPress/gutenberg/blob/5f5b838a315c8044b17781b87b4c1f6cf4fe8a78/packages/block-editor/src/components/block-variation-picker/index.native.js#L55), [bottom-navigation-header](https://github.com/WordPress/gutenberg/blob/5f5b838a315c8044b17781b87b4c1f6cf4fe8a78/packages/components/src/mobile/bottom-sheet/navigation-header.native.js#L102), [segmented-control](https://github.com/WordPress/gutenberg/blob/5f5b838a315c8044b17781b87b4c1f6cf4fe8a78/packages/components/src/mobile/segmented-control/index.native.js#L50), etc) so I thought it'd be good to use it to keep things uniform. WDYT?

dcalhoun replied:
Ah, I see. Thank you for providing that context. I think we should follow the existing pattern as you have here. 

I do wonder if it is limiting the a11y of the product. I hope to research it more at a later time.

dcalhoun commented on packages/editor/src/components/editor-help/index.native.js:
```diff
@@ -0,0 +1,92 @@
+							const key = topic.label;
+							return (
+								<HelpTopicRow
+									key={ key }
+									label={ topic.label }
```
Largely my own curiosity, but what is the rationale for assigning a `key` value earlier while the other values are passed directly to `HelpTopicRow`? I.e. `key={ key }` vs `label={ topic.label}`.

AmandaRiu replied:
The `key` used is `topic.label` and I thought it would help with clarity, but looking at it now I could remove the extra code.

dcalhoun commented on packages/editor/src/components/editor-help/index.native.js:
```diff
@@ -0,0 +1,92 @@
+	const bottomSheetHeaderTitleStyle = getStylesFromColorScheme(
+		styles.bottomSheetHeaderTitle,
+		styles.bottomSheetHeaderTitleDark
+	);
```
What do you think about leveraging the `usePreferredColorScheme` Hook instead of the HoC? My understanding is that this projects generally prefers to Hooks. If we did make this change, we would also need to remove the wrapping HoC and its `import`. 
```suggestion
	const bottomSheetHeaderTitleStyle = usePreferredColorSchemeStyle(
		styles.bottomSheetHeaderTitle,
		styles.bottomSheetHeaderTitleDark
	);
```

dcalhoun commented on packages/editor/src/components/editor-help/index.native.js:
```diff
@@ -0,0 +1,92 @@
+		styles.bottomSheetHeaderTitleDark
+	);
+
+	return useMemo( () => (
```
Largely for my own education, but what is the rationale for leveraging `useMemo` here? Was there a specific performance issue that you observed or just an optimization added in anticipation of issue?

AmandaRiu replied:
@dcalhoun I noticed it's use in other areas of the app when dealing with bottom sheets. Do you think I should remove it?

dcalhoun replied:
@AmandaRiu no, I have no concrete reason to avoid its usage. I asked due to my understanding is that it's generally best to avoid premature optimization. I was curious if there a specific performance issue we needed to address. 

If this pattern is present within other bottom sheets, there may have a very good reason. It could be worthwhile reviewing the history of those other bottom sheets at some point to understand why `useMemo` is present, but I am not requesting that you seek that out right now.

AmandaRiu replied:
I'll definitely research that more during this project and let you know what I find. For now, I'm leaving it as-is.

dcalhoun commented on packages/editor/src/components/editor-help/style.scss:
```diff
@@ -0,0 +1,33 @@
+.sectionContainer {
```
No need to refactor this, but wanted to provide a friendly reminder that new stylesheet files are a good opportunity to explore using the [BEM naming pattern](http://getbem.com/naming/) for selectors while we consider if the practice is a good fit for native stylesheets.

AmandaRiu replied:
Thanks for this information. I'd be interested in how you would change the names of these styles specifically so I can see an example of it applied since there don't seem to be any currently in the native styles code. Thanks!

dcalhoun replied:
Sure thing! I'll put something together and share at a later time.

dcalhoun replied:
Hey @AmandaRiu. 👋🏻 Following up to share an example diff of how these styles might be rewritten following the BEM naming conventions. I hope it's helpful. 

<details><summary>Help BEM Styles</summary>

```diff
diff --git a/packages/editor/src/components/editor-help/index.native.js b/packages/editor/src/components/editor-help/index.native.js
index 3a53c94922..7c3be080a2 100644
--- a/packages/editor/src/components/editor-help/index.native.js
+++ b/packages/editor/src/components/editor-help/index.native.js
@@ -34,21 +34,21 @@ const HELP_TOPICS = [
 
 function EditorHelpTopics( { isVisible, onClose } ) {
 	const bottomSheetHeaderTitleStyle = usePreferredColorSchemeStyle(
-		styles.bottomSheetHeaderTitle,
-		styles.bottomSheetHeaderTitleDark
+		styles.help__heading,
+		styles[ 'help__heading--dark' ]
 	);
 
 	return useMemo( () => (
 		<BottomSheet
 			isVisible={ isVisible }
 			onClose={ onClose }
-			contentStyle={ styles.contentContainer }
+			contentStyle={ styles.help }
 			hideHeader
 			hasNavigation
 		>
 			<BottomSheet.NavigationContainer animate main>
 				<BottomSheet.NavigationScreen name={ __( 'Topics' ) }>
-					<View style={ styles.bottomSheetHeader }>
+					<View style={ styles.help__header }>
 						<Text
 							style={ bottomSheetHeaderTitleStyle }
 							maxFontSizeMultiplier={ 3 }
@@ -56,10 +56,10 @@ function EditorHelpTopics( { isVisible, onClose } ) {
 							{ __( 'How to edit your site' ) }
 						</Text>
 					</View>
-					<View style={ styles.separator } />
+					<View style={ styles.help__separator } />
 					<PanelBody
 						title={ __( 'The basics' ) }
-						style={ styles.sectionContainer }
+						style={ styles.help__section }
 					>
 						{ /* Print out help topics */ }
 						{ HELP_TOPICS.map( ( topic ) => {
diff --git a/packages/editor/src/components/editor-help/style.scss b/packages/editor/src/components/editor-help/style.scss
index 2a0299588b..2c3be45b6e 100644
--- a/packages/editor/src/components/editor-help/style.scss
+++ b/packages/editor/src/components/editor-help/style.scss
@@ -1,22 +1,22 @@
-.contentContainer {
+.help {
 	padding: 0;
 }
 
-.sectionContainer {
+.help__section {
 	padding: 0 16px;
 }
 
-.separator {
+.help__separator {
 	background-color: $light-gray-400;
 	height: 1px;
 	width: 100%;
 }
 
-.separatorDark {
+.help__separator--dark {
 	background-color: $gray-70;
 }
 
-.bottomSheetHeader {
+.help__header {
 	flex-direction: row;
 	align-items: center;
 	padding-right: $grid-unit-20;
@@ -24,7 +24,7 @@
 	min-height: 44px;
 }
 
-.bottomSheetHeaderTitle {
+.help__heading {
 	color: $light-primary;
 	text-align: center;
 	font-weight: 600;
@@ -32,6 +32,6 @@
 	flex: 2;
 }
 
-.bottomSheetHeaderTitleDark {
+.help__heading--dark {
 	color: $dark-primary;
 }
```

</details>

dcalhoun commented on packages/editor/src/components/editor-help/style.scss:
```diff
@@ -0,0 +1,33 @@
+.sectionContainer {
+	padding: 0 0;
```
This has no impact on anything, but I noticed that previous implementations that remove the default padding rely upon the `contentContainerStyle` prop to remove the `ScrollView`'s default padding. I wondered if it is worthwhile to follow that pattern here, rather than removing the padding from `PanelBody`. 

An example of this is block settings [styles](https://github.com/WordPress/gutenberg/blob/b403b977b029911f46247012fa2dcbc42a5aa3cf/packages/block-editor/src/components/block-settings/container.native.scss#L1-L4) modifying the [content styles](https://github.com/WordPress/gutenberg/blob/b403b977b029911f46247012fa2dcbc42a5aa3cf/packages/block-editor/src/components/block-settings/container.native.js#L42) for [the `BottomSheet`](https://github.com/WordPress/gutenberg/blob/b403b977b029911f46247012fa2dcbc42a5aa3cf/packages/components/src/mobile/bottom-sheet/index.native.js#L459).It is likely six of one and half a dozen of another, so I'll let you decide if it is worthwhile. 👍🏻 

Additionally, this can be simplified if desired. 
```suggestion
	padding: 0;
```

AmandaRiu replied:
Thank you for the suggestion 👍

dcalhoun commented on packages/editor/src/components/editor-help/index.native.js:
```diff
@@ -0,0 +1,92 @@
+						} ) }
+					</PanelBody>
+				</BottomSheet.NavigationScreen>
+				{ /* Print out help detail screens */ }
+				{ HELP_TOPICS.map( ( topic ) => {
+					const key = topic.label;
+					return (
+						<HelpDetailNavigationScreen
+							key={ key }
+							name={ topic.label }
+						/>
+					);
+				} ) }
```
We briefly discussed it in chat, but I put together an example of using `BottomSheet.SubSheet`. You can [view the diff here](https://github.com/WordPress/gutenberg/compare/rnmobile/editor-onboarding-basic-help-2...rnmobile/editor-onboarding-basic-help-2-with-subsheet?expand=1), if you'd like to consider leveraging it. I did not heavily test it. 

I also noticed that existing usage of `BottomSheet.SubSheet` outputs an error. It doesn't appear to cause issue in the UI presented, but wanted to note it in case we choose to use `BottomSheet.SubSheet`. I believe [this effect](https://github.com/WordPress/gutenberg/blob/0d34c1c46a49e585e2b7bea1df9821802c4708a8/packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js#L27) needs to be wrapped in `useEffect`, [similar to `BlockMobileToolbar`](https://github.com/WordPress/gutenberg/blob/0d34c1c46a49e585e2b7bea1df9821802c4708a8/packages/block-editor/src/components/block-mobile-toolbar/index.native.js#L56-L62), which was fixed in https://github.com/WordPress/gutenberg/commit/85dc86c3e36f39dab901466388b24d76ec236c71.

<details><summary>SubSheet Error</summary>

```
 ERROR  Warning: Cannot update during an existing state transition (such as within `render`). Render methods should be a pure function of props and state.
    in RCTSafeAreaView (at SafeAreaView.js:51)
    in SafeAreaView (at sub-sheet/index.native.js:24)
    in SlotComponent (at slot.js:91)
    in Slot (at slot-fill/index.native.js:16)
    in Slot (at slot-fill/index.native.js:23)
    in BottomSheetSubSheetSlot (at container.native.js:55)
    in RCTView (at View.js:34)
    in View (at navigation-screen.native.js:86)
    in RCTView (at View.js:34)
    in View (at TouchableHighlight.js:289)
    in TouchableHighlight (at TouchableHighlight.js:352)
    in Unknown (at navigation-screen.native.js:85)
    in RCTScrollContentView (at ScrollView.js:1107)
    in RCTScrollView (at ScrollView.js:1238)
    in ScrollView (at ScrollView.js:1264)
    in ScrollView (at navigation-screen.native.js:84)
    in BottomSheetNavigationScreen (at container.native.js:52)
    in StaticContainer
    in EnsureSingleNavigator (at SceneView.tsx:104)
    in SceneView (at useDescriptors.tsx:150)
    in RCTView (at View.js:34)
    in View (at CardContainer.tsx:198)
    in RCTView (at View.js:34)
    in View (at CardContainer.tsx:197)
    in RCTView (at View.js:34)
    in View (at CardSheet.tsx:33)
    in CardSheet (at Card.tsx:543)
    in RCTView (at View.js:34)
    in View (at createAnimatedComponent.js:217)
    in AnimatedComponent (at createAnimatedComponent.js:278)
    in AnimatedComponentWrapper (at Card.tsx:525)
    in PanGestureHandler (at GestureHandler.native.tsx:13)
    in PanGestureHandler (at Card.tsx:519)
    in RCTView (at View.js:34)
    in View (at createAnimatedComponent.js:217)
    in AnimatedComponent (at createAnimatedComponent.js:278)
    in AnimatedComponentWrapper (at Card.tsx:515)
    in RCTView (at View.js:34)
    in View (at Card.tsx:509)
    in Card (at CardContainer.tsx:167)
    in CardContainer (at CardStack.tsx:552)
    in RCTView (at View.js:34)
    in View (at Screens.tsx:70)
    in MaybeScreen (at CardStack.tsx:545)
    in RCTView (at View.js:34)
    in View (at Screens.tsx:48)
    in MaybeScreenContainer (at CardStack.tsx:455)
    in CardStack (at StackView.tsx:437)
    in KeyboardManager (at StackView.tsx:435)
    in RNCSafeAreaProvider (at SafeAreaContext.tsx:76)
    in SafeAreaProvider (at SafeAreaProviderCompat.tsx:42)
    in SafeAreaProviderCompat (at StackView.tsx:432)
    in RCTView (at View.js:34)
    in View (at StackView.tsx:431)
    in StackView (at createStackNavigator.tsx:82)
    in StackNavigator (at navigation-container.native.js:145)
    in EnsureSingleNavigator (at BaseNavigationContainer.tsx:376)
    in BaseNavigationContainer (at NavigationContainer.tsx:91)
    in ThemeProvider (at NavigationContainer.tsx:90)
    in NavigationContainer (at navigation-container.native.js:144)
    in RCTView (at View.js:34)
    in View (at navigation-container.native.js:132)
    in BottomSheetNavigationContainer (at container.native.js:46)
    in RCTView (at View.js:34)
    in View (at bottom-sheet/index.native.js:541)
    in RCTView (at View.js:34)
    in View (at keyboard-avoiding-view.native.js:118)
    in KeyboardAvoidingView (at bottom-sheet/index.native.js:518)
    in RCTView (at View.js:34)
    in View (at createAnimatedComponent.js:217)
    in AnimatedComponent (at createAnimatedComponent.js:278)
    in AnimatedComponentWrapper (at createAnimatableComponent.js:599)
    in withAnimatable(View) (created by ReactNativeModal)
    in RCTView (at View.js:34)
    in View (at AppContainer.js:107)
    in RCTView (at View.js:34)
    in View (at AppContainer.js:134)
    in AppContainer (at Modal.js:228)
    in RCTView (at View.js:34)
    in View (at Modal.js:249)
    in VirtualizedListContextResetter (at Modal.js:247)
    in RCTModalHostView (at Modal.js:234)
    in Modal (created by ReactNativeModal)
    in ReactNativeModal (at bottom-sheet/index.native.js:489)
    in BottomSheet (at with-preferred-color-scheme/index.native.js:30)
    in WithPreferredColorScheme(BottomSheet) (at container.native.js:38)
    in BottomSheetSettings (at with-dispatch/index.js:96)
    in WithDispatch(BottomSheetSettings) (at with-select/index.js:57)
    in Unknown (at pure/index.tsx:43)
    in WithSelect(WithDispatch(BottomSheetSettings)) (at layout/index.native.js:172)
    in RCTView (at View.js:34)
    in View (at keyboard-avoiding-view/index.ios.js:123)
    in RCTView (at View.js:34)
    in View (at KeyboardAvoidingView.js:209)
    in KeyboardAvoidingView (at createAnimatedComponent.js:217)
    in AnimatedComponent (at createAnimatedComponent.js:278)
    in AnimatedComponentWrapper (at keyboard-avoiding-view/index.ios.js:104)
    in KeyboardAvoidingView (at layout/index.native.js:160)
    in RCTSafeAreaView (at SafeAreaView.js:51)
    in SafeAreaView (at layout/index.native.js:137)
    in RCTView (at View.js:34)
    in View (at tooltip/index.native.js:268)
    in TooltipSlot (at layout/index.native.js:136)
    in Layout (at with-preferred-color-scheme/index.native.js:30)
    in WithPreferredColorScheme(Layout) (at with-select/index.js:57)
    in Unknown (at pure/index.tsx:43)
    in WithSelect(WithPreferredColorScheme(Layout)) (at editor.native.js:167)
    in BlockRefsProvider (at provider/index.native.js:28)
    in BlockEditorProvider (at with-registry-provider.js:24)
    in Unknown (at with-registry/index.js:23)
    in WithRegistryProvider(BlockEditorProvider) (at provider/index.js:108)
    in BlockContextProvider (at provider/index.js:107)
    in EntityProvider (at provider/index.js:106)
    in EntityProvider (at provider/index.js:105)
    in EditorProvider (at with-registry-provider.js:27)
    in Unknown (at with-registry/index.js:23)
    in WithRegistryProvider(EditorProvider) (at provider/index.native.js:305)
    in NativeEditorProvider (at with-dispatch/index.js:96)
    in WithDispatch(NativeEditorProvider) (at with-select/index.js:57)
    in Unknown (at pure/index.tsx:43)
    in WithSelect(WithDispatch(NativeEditorProvider)) (at editor.native.js:160)
    in SlotFillProvider (at editor.native.js:159)
    in Editor (at with-dispatch/index.js:96)
    in WithDispatch(Editor) (at with-select/index.js:57)
    in Unknown (at pure/index.tsx:43)
    in WithSelect(WithDispatch(Editor)) (at src/index.native.js:31)
    in Unknown (at renderApplication.js:47)
    in RCTView (at View.js:34)
    in View (at AppContainer.js:107)
    in RCTView (at View.js:34)
    in View (at AppContainer.js:134)
    in AppContainer (at renderApplication.js:40)
```

</details>

AmandaRiu replied:
Wow, thank you for all this extra work and research you did! I really appreciate it @dcalhoun . I had tried using the SubSheet before but couldn't get it to work for me. I implemented your version of it in 4b1e3b5b8d04dafb5e7a3a718c41f49aab3ffdff. I also fixed the `BottomSheet.SubSheet` render error in 4b1e3b5b8d04dafb5e7a3a718c41f49aab3ffdff. Please let me know what you think 👍

dcalhoun replied:
You're welcome! Happy to do it. I think your new implementation looks great and helps us avoid redundant code for a common pattern of sub-sheets. Thanks!

dcalhoun commented on packages/editor/src/components/editor-help/help-detail-navigation-screen.js:
```diff
@@ -0,0 +1,37 @@
+	};
+
+	return (
+		<BottomSheet.NavigationScreen name={ __( 'What is a block?' ) }>
```
I would think this `name` property needs to be dynamic to match the name of the screen itself, i.e. set to `name`. My understanding is that `name` is the string label used to select _which_ screen to navigate to when calling `navigation.navigate(<screen-name>)`. I am a bit puzzled as to how this is working currently, but it does appear to work as-is. WDYT? 

Note: This point would like become moot if we decide to use `BottomScreen.SubSheet`.

AmandaRiu replied:
I believe it worked because each sheet would have a different name (albeit localized). Doesn't matter at this point though since I did end up using `BottomSheet.SubSheet` thanks to your suggestion 👍

dcalhoun reviewed (commented):
This is looking great! 🎉 I tested on an iPhone 12 mini simulator and Samsung Galaxy S20. The navigation worked as expected, with the correct titles displayed for each section. I left a few comments for consideration, most of them are not blocking and up to your discretion — mostly curious of your opinions. 

One design issue I noticed is that the back button alignment does not appear to match the other instances I found. The images below are from iOS, but the same alignment issue is present for Android as well. 

| iOS Block Settings | iOS Help Sections |
| --- | --- | 
| <img width="401" alt="alignment-color" src="https://user-images.githubusercontent.com/438664/125106381-2fbbe200-e0a5-11eb-8d9e-1d53ee9e353b.png"> | <img width="401" alt="alignment-help" src="https://user-images.githubusercontent.com/438664/125106403-38141d00-e0a5-11eb-9bf1-fbdf5e6f71eb.png"> |

Commit:
Remove unnecessary react-navigation dependency

Commit:
Move imports to appropriate sections

Commit:
Use hook instead of HoC

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>
Use hook for styles instead of HoC

Commit:
Use the ".native.js" file extension for native files

AmandaRiu committed:
Remove unnecessary padding value

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

Commit:
Clean up code

Commit:
Remove unnecessary dependency

Commit:
Fix padding on subsheet

Commit:
Switch to using BottomSheet.SubSheet

Commit:
Update BottomSheet to prevent a render error

AmandaRiu committed:
localize the string Topics

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

AmandaRiu commented:
@dcalhoun Thank you so much for such a thorough review! I've addressed all the comments and fixed the styling in the padding in the subsheet. Ready for another round!

dcalhoun commented on packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js:
```diff
@@ -17,16 +17,24 @@ const BottomSheetSubSheet = ( {
 	isFullScreen,
 } ) => {
+	const BottomSheetScreen = ( { onMount } ) => {
+		useEffect( onMount, [] );
```
💯  Thank you for addressing this error/broken window. ❤️

dcalhoun commented on packages/editor/src/components/editor-help/help-detail-navigation-screen.native.js:
```diff
@@ -0,0 +1,37 @@
+		<BottomSheet.NavigationScreen name={ __( 'What is a block?' ) }>
+			<BottomSheet.NavigationHeader
+				screen={ name }
+				leftButtonOnPress={ goBack }
```
No necessarily requesting a change, but we could simplify this if desired. Merely sharing for future reference. 
```suggestion
				leftButtonOnPress={ navigation.goBack }
```

AmandaRiu replied:
Nice! Thanks for that. I'll apply it in my next PR 👍

dcalhoun reviewed (approved):
LGTM. 💯 Thank you for all your hard work on this. I verified it works as expected on an iPhone 12 mini simulator and Samsung Galaxy S20, including navigation via screen readers.

hypest commented:
Great work here @AmandaRiu! My understanding is that the feature is not really exposed to the end user since it's only accessible via the debug menu atm, right? If so, can you update the PR description to make it more clear that this is not yet available to "production"? Thanks!

AmandaRiu commented:
> Great work here @AmandaRiu! My understanding is that the feature is not really exposed to the end user since it's only accessible via the debug menu atm, right? If so, can you update the PR description to make it more clear that this is not yet available to "production"? Thanks!

Thank you for the suggestion @hypest! I've added an additional note for clarity 👍

jasmussen commented:
@AmandaRiu I'm woefully late to this one, but any reason why this one added a new "alignJustifyAlt" icon to the library? In general when we add to the library we should be careful to only add icons that are used across web and native, and generally go through a design pass first, since these get published as part of the icon library.

In this case there was already an existing justification icon. But judging by the screenshots, this appears to be a custom icon for the onboarding. Can we remove it from the library and just use the vectors directly?

Edit: same with `cogAlt`, `trashFilled` and `sparkles`, though the latter is from a separate PR.

AmandaRiu commented:
Hi @jasmussen!

> @AmandaRiu I'm woefully late to this one, but any reason why this one added a new "alignJustifyAlt" icon to the library? In general when we add to the library we should be careful to only add icons that are used across web and native, and generally go through a design pass first, since these get published as part of the icon library.
> 
> In this case there was already an existing justification icon. But judging by the screenshots, this appears to be a custom icon for the onboarding. Can we remove it from the library and just use the vectors directly?
> 
> Edit: same with `cogAlt`, `trashFilled` and `sparkles`, though the latter is from a separate PR.

I'm not sure honestly, it's been too long since I worked on this feature and i haven't been on this project for a while so I don't remember why those assets were used. If you think they should be swapped out then I'd defer to you or a designer to decide that.

jasmussen commented:
Thanks for the update! I'll see if I can put a PR together to move them to be local at least, then we can discuss on that thread whether bespoke icons are needed at all. Mind if I CC you on the PR? Otherwise, happy to CC someone else?

AmandaRiu commented:
> Thanks for the update! I'll see if I can put a PR together to move them to be local at least, then we can discuss on that thread whether bespoke icons are needed at all. Mind if I CC you on the PR? Otherwise, happy to CC someone else?

Thanks for checking with me @jasmussen! I'm so far removed from working in this repo at this point it may make more sense to have someone else do the PR review.

jasmussen commented:
Followed up in https://github.com/WordPress/gutenberg/pull/38849!
